### PR TITLE
NEW: Feature Stain Normalisation Ruifrok & Reinhard

### PR DESCRIPTION
**Pull request by simongraham**
Added stain normalisation code based on the [StainTools](https://github.com/Peter554/StainTools) library. Methods implemented:

- Reinhard colour normalisation
- Ruifrok & Johnston stain normalisation

Files added/modified:

- `tiatoolbox/tools/stainnorm.py`
- `tiatoolbox/utils/stainnorm.py`
- `tests/test_tiatoolbox.py`
- `tiatoolbox/cli.py`
- `docs/usage.rst`

Requiring attention -> please check the tests. This is done based on storing normalised images in the repo and comparing the results to those images.

Not yet implemented from StainTools:

- Tissue mask method
- Brightness standardiser

Need to discuss whether to add these.

### CODE REVIEW

File: tests/test_tiatoolbox.py
Comment (shaneahmed) [inline tests/test_tiatoolbox.py:10]: Please use 'from tiatoolbox.tools.stainnorm import reinhard_colour_normaliser, stain_normaliser` instead of importing in two lines.

File: tiatoolbox/cli.py
Comment (shaneahmed) [inline tiatoolbox/cli.py:69]: We removed multiproc from develop branch, best to merge latest develop branch into this branch to avoid these issues.

File: tiatoolbox/tools/stainnorm/reinhard_colour_normaliser.py
Comment (shaneahmed) [inline tiatoolbox/tools/stainnorm/reinhard_colour_normaliser.py:61]: ndarray instead of RGB.

File: tiatoolbox/cli.py
Comment (simongraham) [reply tiatoolbox/cli.py]: Fixed in [this commit](https://github.com/TIA-Lab/tiatoolbox/pull/35/commits/47deccd6060278762872c1930d14707625aa2b2d)

File: tests/test_tiatoolbox.py
Comment (simongraham) [reply tests/test_tiatoolbox.py]: Fixed in [this commit](https://github.com/TIA-Lab/tiatoolbox/pull/35/commits/46b2c26c5dc5862f1e4106f2a5347ddb54a59d46)

File: tiatoolbox/tools/stainnorm/reinhard_colour_normaliser.py
Comment (simongraham) [reply tiatoolbox/tools/stainnorm/reinhard_colour_normaliser.py]: Fixed in [this commit](https://github.com/TIA-Lab/tiatoolbox/pull/35/commits/38af1ea1723fb76a8c553e6468a60ddc6c75c999)

File: tests/test_tiatoolbox.py
Comment (shaneahmed) [inline tests/test_tiatoolbox.py:430]: Do we need to include these images? Can we upload them on tia webpage and get access from there?

File: tiatoolbox/cli.py
Comment (shaneahmed) [inline tiatoolbox/cli.py:66]: Did you add a cli function for stain normalisation? This needs to be added and also need to write a test for it.

File: tests/test_tiatoolbox.py
Comment (shaneahmed) [inline tests/test_tiatoolbox.py:436]: Can we define the classes in a single file? e.g., we can have stainnorm.ReinhardColourNormaliser() and stainnorm.Ruifork()?

File: tiatoolbox/tools/stainnorm/reinhard_colour_normaliser.py
Comment (shaneahmed) [inline tiatoolbox/tools/stainnorm/reinhard_colour_normaliser.py:9]: Need to define attributes for this class.

File: tiatoolbox/tools/stainnorm/stain_normaliser.py
Comment (shaneahmed) [inline tiatoolbox/tools/stainnorm/stain_normaliser.py:13]: attributes are not complete

**Review by shaneahmed — CHANGES_REQUESTED** [review]
Overall looks fine to me but I think there is still work which needs to be done. Please can you add the following
- Add attributes in docstring for all the classes. 
- Add command line functions in cli.py
- Update [usage.rst](https://github.com/TIA-Lab/tiatoolbox/blob/feature-stainnorm-ruifrok-reinhard/docs/usage.rst) so the documentation is added to readthedocs.

File: tiatoolbox/cli.py
Comment (simongraham) [reply tiatoolbox/cli.py:66]: No, I didn't add that. I Will do it.

File: tests/test_tiatoolbox.py
Comment (simongraham) [reply tests/test_tiatoolbox.py:430]: Yes this is a good idea and was something I wanted to discuss too. Where exactly can we upload them?

File: tiatoolbox/tools/stainnorm/reinhard_colour_normaliser.py
Comment (simongraham) [reply tiatoolbox/tools/stainnorm/reinhard_colour_normaliser.py:9]: Can you define what you mean by attributes here? Class attributes or instance attributes?

File: tiatoolbox/tools/stainnorm/stain_normaliser.py
Comment (simongraham) [reply tiatoolbox/tools/stainnorm/stain_normaliser.py:13]: Can you define what you mean by attributes here? Class attributes or instance attributes?

File: tests/test_tiatoolbox.py
Comment (simongraham) [reply tests/test_tiatoolbox.py:436]: The way of doing this that I can think of would be to create a script `tiatoolbox/tools/stainnorm.py` and then defining each stain normalisation class within there. Then we can have a stain_utils directory for the utils. What do you think?

Comment (shaneahmed) [reply tests/test_tiatoolbox.py:436]: Yes sounds good to me.

File: tests/test_tiatoolbox.py
Comment (shaneahmed) [reply tests/test_tiatoolbox.py:430]: You can upload [here](https://warwick.ac.uk/fac/sci/dcs/research/tia/tiatoolbox/files)

File: tiatoolbox/tools/stainnorm/reinhard_colour_normaliser.py
Comment (shaneahmed) [reply tiatoolbox/tools/stainnorm/reinhard_colour_normaliser.py:9]: Please consider the example at this [link](https://github.com/TIA-Lab/tiatoolbox/blob/a826f658dd02671631d83cfc42676b2858548fb5/tiatoolbox/dataloader/wsireader.py#L35)

File: tiatoolbox/tools/stainnorm/stain_normaliser.py
Comment (shaneahmed) [reply tiatoolbox/tools/stainnorm/stain_normaliser.py:13]: Please consider the example at this [link](https://github.com/TIA-Lab/tiatoolbox/blob/a826f658dd02671631d83cfc42676b2858548fb5/tiatoolbox/dataloader/wsireader.py#L35)

File: tests/test_tiatoolbox.py
Comment (simongraham) [reply tests/test_tiatoolbox.py]: Modified in [this commit](https://github.com/TIA-Lab/tiatoolbox/pull/35/commits/ad1209a509e700c326ff847fa61437b264294582).

File: tiatoolbox/tools/stainnorm/reinhard_colour_normaliser.py
Comment (simongraham) [reply tiatoolbox/tools/stainnorm/reinhard_colour_normaliser.py]: This isn't really clear by looking at that example. Are you referring to the variables defined as arguments in the __init__ or the instance variables defined in __init__ itself? Here, I didn't include attributes because __init__ only takes self as an argument.

Comment (shaneahmed) [reply tiatoolbox/tools/stainnorm/reinhard_colour_normaliser.py]: self.target_means = means
self.target_stds = stds

File: tiatoolbox/tools/stainnorm/stain_normaliser.py
Comment (shaneahmed) [reply tiatoolbox/tools/stainnorm/stain_normaliser.py]: self.target_means = means
self.target_stds = stds

target_means and target_stds will be class attributes?

File: tests/test_tiatoolbox.py
Comment (simongraham) [reply tests/test_tiatoolbox.py]: Addressed in [this commit](https://github.com/TIA-Lab/tiatoolbox/pull/35/commits/3cfb8fe5f2b77998950484dfed6bd504af4670dd).

File: docs/usage.rst
Comment (shaneahmed) [inline docs/usage.rst:74]: Please can we move Tools before Utils? Just following alphabetical order.

File: tiatoolbox/utils/exceptions.py
Comment (shaneahmed) [inline tiatoolbox/utils/exceptions.py:32]: Can you write a quick test for this code for coverage?

File: tiatoolbox/tools/stainnorm.py
Comment (shaneahmed) [inline tiatoolbox/tools/stainnorm.py:31]: Can we not call it RuiforkStainNormaliser?

File: tiatoolbox/tools/stainnorm.py
Comment (shaneahmed) [inline tiatoolbox/tools/stainnorm.py:30]: Shall we define an ABC?

Comment (simongraham) [reply tiatoolbox/tools/stainnorm.py:30]: We define ABC for the stain extraction method. Check `tiatoolbox/utils/stainnorm.py`

File: tiatoolbox/tools/stainnorm.py
Comment (simongraham) [reply tiatoolbox/tools/stainnorm.py:31]: This is a design choice. With the current setup, all that we would need to do in this class is change the `self.extractor`, which is set by the `method` argument. If we do the above, then we will need to define a completely new class.

File: tiatoolbox/tools/stainnorm.py
Comment (John-P) [inline tiatoolbox/tools/stainnorm.py:17]: The company name on this line needs to be changed. We are not the Blender Foundation.

File: tiatoolbox/utils/stainnorm.py
Comment (shaneahmed) [inline tiatoolbox/utils/stainnorm.py:49]: Allow the user to input stain matrix.

**Review by John-P — CHANGES_REQUESTED** [review]
Consider moving stain extraction to a file alongside stainnorm maybe stainextract? Need to think about how we will add more fixed stain matrices for Ruifrok method. Also, are we using Ruifrok to refer to the matrix or the method or both?

File: docs/usage.rst
Comment (simongraham) [reply docs/usage.rst]: Resolved

File: tiatoolbox/tools/stainnorm.py
Comment (simongraham) [reply tiatoolbox/tools/stainnorm.py]: Fixed in [this commit](https://github.com/TIA-Lab/tiatoolbox/commit/f823ffaac09f575459a9ad590151e1ef7f0bd00e)

File: tiatoolbox/utils/stainnorm.py
Comment (simongraham) [reply tiatoolbox/utils/stainnorm.py]: Added CustomNormaliser that allows a user to input their own stain matrix

File: tiatoolbox/utils/exceptions.py
Comment (simongraham) [reply tiatoolbox/utils/exceptions.py:32]: Added in[ this commit](https://github.com/TIA-Lab/tiatoolbox/pull/35/commits/a5d3d7f6097a144f92d1e5b1cc183ed1464bfbfd).

**Comment by simongraham** [comment]
Following review, changes made:

- Moved stainnorm_utils.py from `utils` directory and renamed to `stainextract.py`
- Restructured the code
- Add helper function `get_normaliser()` to enable smooth usage between methods.
- Add `MethodNotSupported` test
- Update `usage.rst`

File: tests/test_tiatoolbox.py
Comment (shaneahmed) [inline tests/test_tiatoolbox.py:419]: I like it!

File: tiatoolbox/cli.py
Comment (shaneahmed) [inline tiatoolbox/cli.py:294]: For custom method, do you need to pass stain matrix?

File: tiatoolbox/tools/stainnorm.py
Comment (shaneahmed) [inline tiatoolbox/tools/stainnorm.py:34]: How about extractor?

**Review by shaneahmed — APPROVED** [review]
Thanks Simon. It looks great, we can add Vahadane and Macenko next. Please address the comments I have added and once you get the second review we should be ready to merge.

File: tiatoolbox/tools/stainnorm.py
Comment (simongraham) [reply tiatoolbox/tools/stainnorm.py:35]: Yes, I missed this. Good spot. Pushing and will link the commit.

File: tiatoolbox/cli.py
Comment (simongraham) [reply tiatoolbox/cli.py:294]: Yes - will add stain matrix as argument and default to `None` for other methods.

Comment (shaneahmed) [reply tiatoolbox/cli.py:294]: which argument will be used for passing stain matrix?

File: tiatoolbox/tools/stainnorm.py
Comment (shaneahmed) [inline tiatoolbox/tools/stainnorm.py:34]: It's not added properly. it should be `extractor (type) : method specific ...`

File: tiatoolbox/cli.py
Comment (simongraham) [reply tiatoolbox/cli.py:294]: https://github.com/TIA-Lab/tiatoolbox/blob/feature-stainnorm-ruifrok-reinhard/tiatoolbox/cli.py#L298

File: tiatoolbox/tools/stainnorm.py
Comment (simongraham) [reply tiatoolbox/tools/stainnorm.py:34]: okay - this will be an object. Would (obj) be the right type to use here?

Comment (shaneahmed) [reply tiatoolbox/tools/stainnorm.py:34]: `extractor (CustomExtractor, RuifrokExtractor) : method specific ...`

File: tests/test_tiatoolbox.py
Comment (John-P) [inline tests/test_tiatoolbox.py:3]: Very minor point. I thought 'Tests' was fine before. Starting the sentence with pytests sounds weird to me. :shrug:

File: tests/test_tiatoolbox.py
Comment (John-P) [inline tests/test_tiatoolbox.py:147]: Since the target image is small, should be not included it as a recourse in the pip package so that it is included with the python code. Just seems a bit silly to download it each time. Especially as it will be used outside of unit tests also.

File: tiatoolbox/cli.py
Comment (shaneahmed) [reply tiatoolbox/cli.py:294]: ok great. Please can you add an example how the matrix can be passed as input? I don't think string will work here.

File: tiatoolbox/tools/__init__.py
Comment (John-P) [inline tiatoolbox/tools/__init__.py:2]: I know it's a very small file but missing the GPL header here.

**Review by John-P — CHANGES_REQUESTED** [review]
Mostly just minor things such as wording and formatting plus one missing GPL header. Otherwise, looks good.

**Review by John-P — APPROVED** [review]

File: tests/test_tiatoolbox.py
Comment (shaneahmed) [reply tests/test_tiatoolbox.py:3]: All we need is to be consistent, whether Tests or pytests

Comment (simongraham) [reply tests/test_tiatoolbox.py:3]: This is done for most strings in this file. Shall we change all to Test?

File: tests/test_tiatoolbox.py
Comment (simongraham) [reply tests/test_tiatoolbox.py:147]: It's up to us I guess. I did it the way you are describing initially and then changed to how it is structured now. Assume that the user will want to supply their own target image? Happy to add an example target image instead within the repo. What do we think?

File: tests/test_tiatoolbox.py
Comment (simongraham) [reply tests/test_tiatoolbox.py:3]: I will change all to `Tests` now

File: tiatoolbox/tools/stainnorm.py
Comment (simongraham) [reply tiatoolbox/tools/stainnorm.py]: okay - fixed

File: tests/test_tiatoolbox.py
Comment (simongraham) [reply tests/test_tiatoolbox.py:3]: Fixed [here](https://github.com/TIA-Lab/tiatoolbox/pull/35/commits/0f1081053337cdbaabe8205a779109c9e04d3f81).

File: tiatoolbox/tools/__init__.py
Comment (simongraham) [reply tiatoolbox/tools/__init__.py:22]: This was actually the case with **all** `__init__.py` files. Changed in [this](https://github.com/TIA-Lab/tiatoolbox/pull/35/commits/593da318650b992eee629b982adc3b0f4631a744) commit.

File: tiatoolbox/tools/stainnorm.py
Comment (shaneahmed) [reply tiatoolbox/tools/stainnorm.py]: In this commit https://github.com/TIA-Lab/tiatoolbox/commit/57aef63606641ae0c3c15f11383380d944c1b4ba

**Pull request merged**
